## Starbucks Offers Deep Analysis

In [20]:
#imports 

import numpy as np
import pandas as pd
import math 
import json 
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.preprocessing import OneHotEncoder

### Data Exploration

In [40]:
#Reading in JSONs
portfolio = pd.read_json('portfolio.json', orient='records', lines=True)
profile = pd.read_json('profile.json', orient='records', lines=1)
transcript = pd.read_json('transcript.json', orient='records', lines=True)

In [41]:
print('There are {} rows and {} columns in the portfolio dataframe.'.format(portfolio.shape[0], portfolio.shape[1]))

portfolio.describe(include='all')

There are 10 rows and 6 columns in the portfolio dataframe.


reward                      channels  difficulty   duration  \
count   10.000000                            10   10.000000  10.000000   
unique        NaN                             4         NaN        NaN   
top           NaN  [web, email, mobile, social]         NaN        NaN   
freq          NaN                             4         NaN        NaN   
mean     4.200000                           NaN    7.700000   6.500000   
std      3.583915                           NaN    5.831905   2.321398   
min      0.000000                           NaN    0.000000   3.000000   
25%      2.000000                           NaN    5.000000   5.000000   
50%      4.000000                           NaN    8.500000   7.000000   
75%      5.000000                           NaN   10.000000   7.000000   
max     10.000000                           NaN   20.000000  10.000000   

       offer_type                                id  
count          10                                10  
unique          3                                10  
top          bogo  ae264e3637204a6fb9bb56bc8210ddfd  
freq            4                                 1  
mean          NaN                               NaN  
std           NaN                               NaN  
min           NaN                               NaN  
25%           NaN                               NaN  
50%           NaN                               NaN  
75%           NaN                               NaN  
max           NaN                               NaN

In [23]:
portfolio.head()

reward                      channels  difficulty  duration     offer_type  \
0      10       [email, mobile, social]          10         7           bogo   
1      10  [web, email, mobile, social]          10         5           bogo   
2       0          [web, email, mobile]           0         4  informational   
3       5          [web, email, mobile]           5         7           bogo   
4       5                  [web, email]          20        10       discount   

                                 id  
0  ae264e3637204a6fb9bb56bc8210ddfd  
1  4d5c57ea9a6940dd891ad53e9dbe8da0  
2  3f207df678b143eea3cee63160fa8bed  
3  9b98b8c7a33c4b65b9aebfe6a799e6d9  
4  0b1e1539f2cc45b7b9fa7c272da2e1d7

In [24]:
print('There are {} rows and {} columns in the profile dataframe.'.format(profile.shape[0], profile.shape[1]))

profile.describe(include='all')

There are 17000 rows and 5 columns in the profile dataframe.


gender           age                                id  \
count   14825  17000.000000                             17000   
unique      3           NaN                             17000   
top         M           NaN  68be06ca386d4c31939f3a4f0e3dd783   
freq     8484           NaN                                 1   
mean      NaN     62.531412                               NaN   
std       NaN     26.738580                               NaN   
min       NaN     18.000000                               NaN   
25%       NaN     45.000000                               NaN   
50%       NaN     58.000000                               NaN   
75%       NaN     73.000000                               NaN   
max       NaN    118.000000                               NaN   

        became_member_on         income  
count       1.700000e+04   14825.000000  
unique               NaN            NaN  
top                  NaN            NaN  
freq                 NaN            NaN  
mean        2.016703e+07   65404.991568  
std         1.167750e+04   21598.299410  
min         2.013073e+07   30000.000000  
25%         2.016053e+07   49000.000000  
50%         2.017080e+07   64000.000000  
75%         2.017123e+07   80000.000000  
max         2.018073e+07  120000.000000

In [25]:
profile.head()

gender  age                                id  became_member_on    income
0   None  118  68be06ca386d4c31939f3a4f0e3dd783          20170212       NaN
1      F   55  0610b486422d4921ae7d2bf64640c50b          20170715  112000.0
2   None  118  38fe809add3b4fcf9315a9694bb96ff5          20180712       NaN
3      F   75  78afa995795e4d85b5d9ceeca43f5fef          20170509  100000.0
4   None  118  a03223e636434f42ac4c3df47e8bac43          20170804       NaN

In [26]:
print('There are {} rows and {} columns in the transcript dataframe.'.format(transcript.shape[0], transcript.shape[1]))

transcript.describe(include='all')

There are 306534 rows and 4 columns in the transcript dataframe.


person        event  \
count                             306534       306534   
unique                             17000            4   
top     94de646f7b6041228ca7dec82adb97d2  transaction   
freq                                  51       138953   
mean                                 NaN          NaN   
std                                  NaN          NaN   
min                                  NaN          NaN   
25%                                  NaN          NaN   
50%                                  NaN          NaN   
75%                                  NaN          NaN   
max                                  NaN          NaN   

                                                   value           time  
count                                             306534  306534.000000  
unique                                              5121            NaN  
top     {'offer id': '2298d6c36e964ae4a3e7e9706d1fb8c2'}            NaN  
freq                                               14983            NaN  
mean                                                 NaN     366.382940  
std                                                  NaN     200.326314  
min                                                  NaN       0.000000  
25%                                                  NaN     186.000000  
50%                                                  NaN     408.000000  
75%                                                  NaN     528.000000  
max                                                  NaN     714.000000

In [27]:
transcript.head()

person           event  \
0  78afa995795e4d85b5d9ceeca43f5fef  offer received   
1  a03223e636434f42ac4c3df47e8bac43  offer received   
2  e2127556f4f64592b11af22de27a7932  offer received   
3  8ec6ce2a7e7949b1bf142def7d0e0586  offer received   
4  68617ca6246f4fbc85e91a2a49552598  offer received   

                                              value  time  
0  {'offer id': '9b98b8c7a33c4b65b9aebfe6a799e6d9'}     0  
1  {'offer id': '0b1e1539f2cc45b7b9fa7c272da2e1d7'}     0  
2  {'offer id': '2906b810c7d4411798c6938adc9daaa5'}     0  
3  {'offer id': 'fafdcd668e3743c1bb461111dcafc2a4'}     0  
4  {'offer id': '4d5c57ea9a6940dd891ad53e9dbe8da0'}     0

#### Handling Missing Values

In [28]:
portfolio.isnull().sum()

reward        0
channels      0
difficulty    0
duration      0
offer_type    0
id            0
dtype: int64

In [29]:
transcript.isnull().sum()

person    0
event     0
value     0
time      0
dtype: int64

In [30]:
profile.isnull().sum()

gender              2175
age                    0
id                     0
became_member_on       0
income              2175
dtype: int64

There are 2175 missing values in the "gender" column of the profile dataframe, which is a relatively small % of the dataset, so I'll replace the missing values with the column's mode. There are also the same amount of missing values in the "income" column of that dataframe, which I'll fill with the mean of the column's values. 

In [42]:
profile['gender'] = profile['gender'].fillna(profile['gender'].mode()[0])

In [43]:
profile['income'] = profile['income'].fillna(profile['income'].mean())

In [44]:
profile.isnull().sum()

gender              0
age                 0
id                  0
became_member_on    0
income              0
dtype: int64

In [45]:
#Splitting channels column
channel_list = []

for channel in portfolio['channels']:
    for i in channel:
        if i not in channel_list:
            channel_list.append(i)
            
for channel in channel_list : 
    portfolio[channel] = portfolio['channels'].apply(lambda x:
                                                     1 if channel in x else 0)
    
portfolio.drop('channels', axis = 1, inplace = True)

In [46]:
portfolio

reward  difficulty  duration     offer_type  \
0      10          10         7           bogo   
1      10          10         5           bogo   
2       0           0         4  informational   
3       5           5         7           bogo   
4       5          20        10       discount   
5       3           7         7       discount   
6       2          10        10       discount   
7       0           0         3  informational   
8       5           5         5           bogo   
9       2          10         7       discount   

                                 id  email  mobile  social  web  
0  ae264e3637204a6fb9bb56bc8210ddfd      1       1       1    0  
1  4d5c57ea9a6940dd891ad53e9dbe8da0      1       1       1    1  
2  3f207df678b143eea3cee63160fa8bed      1       1       0    1  
3  9b98b8c7a33c4b65b9aebfe6a799e6d9      1       1       0    1  
4  0b1e1539f2cc45b7b9fa7c272da2e1d7      1       0       0    1  
5  2298d6c36e964ae4a3e7e9706d1fb8c2      1       1       1    1  
6  fafdcd668e3743c1bb461111dcafc2a4      1       1       1    1  
7  5a8bc65990b245e5a138643cd4eb9837      1       1       1    0  
8  f19421c1d4aa40978ebb69ca19b0e20d      1       1       1    1  
9  2906b810c7d4411798c6938adc9daaa5      1       1       0    1

In [47]:
portfolio['offer_type'].unique()

























































array(['bogo', 'informational', 'discount'], dtype=object)